apps/yggdrasil/lib/yggdrasil/user.ex

```diff
@@ -29,14 +37,120 @@ defmodule Yggdrasil.User do
+
+    permissions = query
+    |> Repo.all()
+    |> Enum.flat_map(fn ur -> ur.role.role_permissions end)
+    |> Enum.group_by(fn rp -> rp.resource end)
+    |> Enum.map(fn {r, rps} -> {String.to_atom(r), Enum.map(rps, fn rp -> String.to_atom(rp.permission) end)} end)
+    |> Enum.into(%{})
+
+    %{ user | permissions: permissions}
+  end
+
+  @doc ~S"""
+  Checks the list of resource/permissions pair given against what the user
+  has been granted and returns true if and only if all permissions supplied
+  have been granted for the user.
+
+  ## roles
+  A user can have one or more roles with each role having a set of resources
+  and a set of resources and those resources having a set of permissions. Roles
+  are not checked directly and only serve as a grouping mechanism for resources.
+
+  Since the user can have multiple roles, the permission set checked for a given
+  resource is the unique set across all the roles for that resource.
+
+  So if the user has `Role1` and `Role2` defined as follows:
+
+  * Role1
+    * foo
+      * :read
+      * :write
+  * Role2
+    * foo
+      * :read
+      * :all
+
+  The set of permissions checked for foo will be `[:read, :write, :all]`
+
+  ## Examples
+      # checking a single resource
+      User.is_granted? user, foo: [:read]
+
+      # checking multiple resources is allowed as well
+      User.is_granted? user, foo: [:read], bar: [:read, :write]
   """
-  def hash_password(changeset = %{:valid? => false}) do
+  def is_granted!(user, res_perms) do
+    case is_granted?(user, res_perms) do
+      {:ok, result}  -> result
+      {:error, msgs} -> raise RuntimeError, inspect(msgs)
+    end
+  end
+
+  def is_granted?(_user, []),  do: build_error ["resource keyword list is empty"]
+  def is_granted?(_user, nil), do: build_error ["resource keyword list is nil"]
+  def is_granted?(user, res_perms) do
+    res_perms
+    |> Enum.map(fn {res, perms} -> {res, user.permissions[res], perms} end)
+    |> Enum.map(&has_permissions/1)
+    |> Enum.reduce({:ok, true}, fn
+      {:ok, _},  acc={:ok, false}   -> acc
+      {:ok, res},    {:ok, true}    -> {:ok, res}
+      {:ok, _},  acc={:error, _}    -> acc
+      {:error, msg}, {:ok, _}       -> {:error, [msg]}
+      {:error, msg}, {:error, msgs} -> {:error, [msg|msgs]}
+    end)
+  end
+
+  defp has_permissions({_res, nil, _requried}), do: {:ok, false}
+  defp has_permissions({res, _given, nil}),     do: build_error "required permission list is nil for #{res}"
+  defp has_permissions({res, [], _required}),   do: build_error "given permission list is empty for #{res}"
+  defp has_permissions({res, _given, []}),      do: build_error "required permission list is empty for #{res}"
+  defp has_permissions({_res, given, required}) do
+    given = MapSet.new given
+    required = MapSet.new required
+
+    {:ok, MapSet.subset?(required, given)}
+  end
+
+  defp build_error(msg) do
+    Logger.error msg
+    {:error, msg}
+  end
```

Comment:
I'm not a big fan of only returning an error message, it doesn't provide any data for the caller to make decisions with. I do see the argument of maybe returning `:allowed | { :denied, ... } | { :error, ... }`.

I guess I just don't see why you can't have it all with this approach. Very easily could have an `is_granted?` that returns `true | false` and raises if there were errors, and could even have an `is_granted!` that returns `true` or raises. Certainly there is a valid argument for YAGNI since this isn't a library to be released separately. If you want to only provide the `is_granted?` that seems reasonable.

I'm not sure what is the most idiomatic method. I see that Ecto's [delete/2](https://hexdocs.pm/ecto/Ecto.Repo.html#c:delete/2) has a couple of variations. I'm going to guess if the connection was closed or server unreachable that would be an exception.

I believe a 500 response should include an explanation. Although most cases it'll be a vague "server error occurred".  A 401 is that you require the user to authenticate first, a 403 is that they are authenticated but they are not authorized.